nixos/modules/services/search/elasticsearch.nix

```diff
@@ -163,6 +175,8 @@ in
         User = "elasticsearch";
         PermissionsStartOnly = true;
         LimitNOFILE = "1024000";
+        Restart = "always";
+        TimeoutStartSec = "infinity";
```

Comment:
Can we set this to something like 2 or 5 or 10 minutes instead forever?

Reply:
If we set an arbitrary value here, perfectly functional clusters will fail to start. I think the more proper solution here is to make ES a "notify" type service like on other distros. I'll try to make those changes and update this PR.
https://github.com/elastic/elasticsearch/blob/master/distribution/packages/src/common/systemd/elasticsearch.service

That being said, setting arbitrary timeouts isn't the right thing to do here. It causes unnecessary confusion for anyone that has large/slow nodes. For example, some nodes I manage have startup times of over 15 minutes. Someone else might have to deal with even slower nodes. Short of converting to notify, a timeout of infinity most closely resembles the behavior of elasticsearch in upstream packages.

Reply:
I did some more digging, the [class](https://github.com/elastic/elasticsearch/blob/7f81877ce0ca6b2995bdb988b73609fdee0f82d6/modules/systemd/src/main/java/org/elasticsearch/systemd/SystemdPlugin.java) that makes systemd notify wok is not present in the tarball release of elasticsearch. Looks like the best way to deal with this is to set TimeoutStartSec to infinity, as done in this PR.

Reply:
What about 1h, 6h or 24h? Infinity is infinite to long.

Reply:
Why do you insist on an arbitrary limit here?

Reply:
> Also, infinite startup timeouts are not unprecedented in nixpkgs:

Except maybe clout-init those are also not great choices.

> If infinity being too long is just a matter of your personal taste

it is the essence of infinity that it could take to the end of time. You want predictable and reliable service management. If it normally takes 15 minutes to boot up, you can set the timeout to 30 or 60 minutes if you want to be sure but after that you want to check why it is taking so long, instead of waiting another X hours after someone noticed. So it is a really poor choice to have the default at infinity if 99% of the time it is done in 15 minutes or less because people do not change defaults and are then surprised why the default is not something reasonable which safe guards you against stupid mistakes.

And it is not pointless to kill the elasticsearch server at some point because if it is already taking two days to start I do not think it will start at all and someone should investigate why it is taking so much longer than normal. Maybe the disk is slowly dying, network degradation or the server has to much load and starves to death.

cloud-init could be an exception in the above but I believe that the other services should also have a lower timeout except maybe if they handle it very well themselves. Then we have the question why they do it in the first place, if systemd can already do it.

I mean we can merge it like this but I don't think it is a good idea and choice.

Reply:
I agree with you that finite timeouts make sense for a lot of services, especially when they're just a minor component on a machine. But typically, you would run Elasticsearch on dedicated nodes.

> If it normally takes 15 minutes to boot up, you can set the timeout to 30 or 60 minutes if you want to be sure but after that you want to check why it is taking so long, instead of waiting another X hours after someone noticed.

If you were to monitor the health of your ES cluster through, say, the prometheus node exporter, you would add a check for whether the elasticsearch service is in "active" state. The way this module is set up, you will know if ES is not up. If the ES process gets stuck indefinitely during startup, a restart probably won't fix the issue. For someone trying to investigate the issue, it's far more helpful to have the affected process stay the way it is.

> if it is already taking two days to start I do not think it will start at all and someone should investigate why it is taking so much longer than normal

Agreed, unusually long startup times need to be investigated. But trying to bring attention to unusually long startup times by killing the service rather than simply looking at the unusually long startup time is... unusual.

I understand where you're coming from, but I disagree with you on how to handle a slow/hung elasticsearch instance. Restarting a slow process isn't "predictable and reliable service management". I'll add an option to change the start timeout, but it ought to default to infinity.

Reply:
Maybe we can make an exception here but I would prefer if not more services set such high timeouts.